File: threads/src/main/java/com/sequenia/threads/adapters/FilesAndMediaAdapter.java
package com.sequenia.threads.adapters;

import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewGroup;

import com.sequenia.threads.holders.FileAndMediaViewHolder;
import com.sequenia.threads.holders.FilesDateStampHolder;
import com.sequenia.threads.model.DateRow;
import com.sequenia.threads.model.FileAndMediaItem;
import com.sequenia.threads.model.FileDescription;
import com.sequenia.threads.model.MediaAndFileItem;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

/**
 * Created by yuri on 01.07.2016.
 */
public class FilesAndMediaAdapter extends RecyclerView.Adapter {
    private static final String TAG = "FilesAndMediaAdapter ";
    private static final int TYPE_DATE_ROW = 1;
    private static final int TYPE_FILE_AND_MEDIA_ROW = 2;
    private OnFileClick mOnFileClick;

    private ArrayList<MediaAndFileItem> list;

    public FilesAndMediaAdapter(List<FileDescription> list, OnFileClick onFileClick) {
        this.list = new ArrayList<>();
        mOnFileClick = onFileClick;
        if (list.size() == 0) return;
        this.list.add(new DateRow(list.get(0).getTimeStamp()));
        this.list.add(new FileAndMediaItem(list.get(0)));
        Calendar current = Calendar.getInstance();
        Calendar prev = Calendar.getInstance();

        for (int i = 1; i < list.size(); i++) {
            current.setTimeInMillis(list.get(i).getTimeStamp());
            prev.setTimeInMillis(list.get(i - 1).getTimeStamp());
            this.list.add(new FileAndMediaItem(list.get(i)));
            if (current.get(Calendar.DAY_OF_YEAR) != prev.get(Calendar.DAY_OF_YEAR)) {
                this.list.add(new DateRow(list.get(i).getTimeStamp()));
            }
        }
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        if (viewType == TYPE_DATE_ROW) return new FilesDateStampHolder(parent);
        if (viewType == TYPE_FILE_AND_MEDIA_ROW) return new FileAndMediaViewHolder(parent);
        throw new IllegalStateException("unknown view type");
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, final int position) {
        if (getItemViewType(position) == TYPE_DATE_ROW) {
            FilesDateStampHolder h = (FilesDateStampHolder) holder;
            h.onBind(list.get(position).getTimeStamp());
        }
        if (getItemViewType(position) == TYPE_FILE_AND_MEDIA_ROW) {
            final FileAndMediaViewHolder h = (FileAndMediaViewHolder) holder;
            FileAndMediaItem item = (FileAndMediaItem) list.get(position);
            h.onBind(item.getFileDescription()
                    , new View.OnClickListener() {
                        @Override
                        public void onClick(View v) {
                            if (null != mOnFileClick) {
                                FileAndMediaItem item = (FileAndMediaItem) list.get(h.getAdapterPosition());
                                mOnFileClick.onFileClick(item.getFileDescription());
                            }
                        }
                    });
        }
    }

    @Override
    public int getItemCount() {
        return list.size();
    }

    @Override
    public int getItemViewType(int position) {
        if (list.get(position) instanceof DateRow) return TYPE_DATE_ROW;
        if (list.get(position) instanceof FileAndMediaItem) return TYPE_FILE_AND_MEDIA_ROW;
        return super.getItemViewType(position);
    }

    public interface OnFileClick {
        void onFileClick(FileDescription fileDescription);
    }
}

File: src/main/java/im/threads/ui/adapters/filesAndMedia/FilesAndMediaAdapter.kt
package im.threads.ui.adapters.filesAndMedia

import android.view.ViewGroup
import androidx.core.util.ObjectsCompat
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import im.threads.business.models.DateRow
import im.threads.business.models.FileAndMediaItem
import im.threads.business.models.FileDescription
import im.threads.business.models.MediaAndFileItem
import im.threads.business.utils.FileUtils.getFileName
import im.threads.ui.holders.EmptyViewHolder
import im.threads.ui.holders.FileAndMediaViewHolder
import im.threads.ui.holders.FilesDateStampHolder
import java.util.Calendar
import java.util.Locale

/** Адаптер для списка файлов, отправленных или полученных в чате */
@Suppress("NAME_SHADOWING")
internal class FilesAndMediaAdapter(
    filesList: List<FileDescription?>,
    private val onFileClick: OnFileClick,
    private val onDownloadFileClick: OnFileClick
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    private var list = ArrayList<MediaAndFileItem>()
    private var backup = ArrayList<MediaAndFileItem>()

    init {
        addItems(filesList)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return when (viewType) {
            TYPE_DATE_ROW -> FilesDateStampHolder(parent)
            TYPE_FILE_AND_MEDIA_ROW -> FileAndMediaViewHolder(parent)
            else -> EmptyViewHolder(parent)
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (getItemViewType(position) == TYPE_DATE_ROW) {
            val dateStampHolder = holder as FilesDateStampHolder
            dateStampHolder.onBind(list[position].timeStamp)
        }
        if (getItemViewType(position) == TYPE_FILE_AND_MEDIA_ROW) {
            val fileAndMediaHolder = holder as FileAndMediaViewHolder
            val (fileDescription) = list[position] as FileAndMediaItem
            fileAndMediaHolder.onBind(
                fileDescription,
                {
                    val (fileItem) = list[fileAndMediaHolder.adapterPosition] as FileAndMediaItem
                    onFileClick.onFileClick(fileItem)
                }
            ) {
                val (fileItem) = list[fileAndMediaHolder.adapterPosition] as FileAndMediaItem
                onDownloadFileClick.onDownloadFileClick(fileItem)
            }
        }
    }

    override fun getItemCount(): Int {
        return list.size
    }

    override fun getItemViewType(position: Int): Int {
        return if (list[position] is DateRow) {
            TYPE_DATE_ROW
        } else if (list[position] is FileAndMediaItem) {
            TYPE_FILE_AND_MEDIA_ROW
        } else {
            super.getItemViewType(position)
        }
    }

    /**
     * Проверяет количество элементов в списке на пустоту
     */
    fun isNotEmpty() = itemCount > 0

    /** Очищает текущий список, предварительно создавая его бэкап */
    fun backupAndClear() {
        backup = ArrayList(list)
        updateWithDiffUtil { list.clear() }
    }

    /**
     * Фильтрует список по заданному параметру
     *
     * @param filter строка для фильтрации
     */
    inline fun filter(filter: String?, onFiltered: () -> Unit) {
        var filter = filter
        if (filter == null) filter = ""
        val filteredItems = ArrayList<FileDescription>()
        for (item in backup) {
            if (item is FileAndMediaItem) {
                val (fileDescription, fileName) = item
                val lastPathSegment = fileDescription.fileUri?.lastPathSegment
                val lowercaseLastPath = lastPathSegment?.lowercase(Locale.getDefault())
                val lowercaseFilter = filter.lowercase(Locale.getDefault())
                val isContainsThisLocale = lowercaseLastPath?.contains(lowercaseFilter) == true

                if (isContainsThisLocale) {
                    filteredItems.add(fileDescription)
                } else if (fileDescription.incomingName != null) {
                    val lowercaseName = fileDescription.incomingName?.lowercase(Locale.getDefault())
                    if (lowercaseName?.contains(lowercaseFilter) == true) {
                        filteredItems.add(fileDescription)
                    }
                } else if (fileName.lowercase(Locale.getDefault()).contains(lowercaseFilter)) {
                    filteredItems.add(fileDescription)
                }
            }
        }

        updateWithDiffUtil {
            list = getMediaAndFileItems(filteredItems)
        }
        onFiltered()
    }

    /** Восстанавливает список из бэкапа */
    fun undoClear() {
        updateWithDiffUtil { list = ArrayList(backup) }
        backup.clear()
    }

    /**
     * Обновляет прогресс загрузки для файла
     *
     * @param fileDescription характеристики файла для обновление
     *     прогресса
     */
    fun updateProgress(fileDescription: FileDescription?) {
        for (i in list.indices) {
            if (list[i] is FileAndMediaItem) {
                val mediaItem = list[i] as FileAndMediaItem
                if (ObjectsCompat.equals(mediaItem.fileDescription, fileDescription)) {
                    val itemCopy = mediaItem.copy(
                        fileDescription = fileDescription!!,
                        fileName = mediaItem.fileName
                    )
                    list[i] = itemCopy
                    notifyItemChanged(i)
                }
            }
        }
    }

    /**
     * Показывает ошибку загрузки для файла
     *
     * @param fileDescription характеристики файла для отображения
     *     ошибки
     */
    fun onDownloadError(fileDescription: FileDescription?) {
        for (i in list.indices) {
            if (list[i] is FileAndMediaItem) {
                val (fileDescription1) = list[i] as FileAndMediaItem
                val itemViewType = getItemViewType(i)
                if (ObjectsCompat.equals(fileDescription1, fileDescription) &&
                    itemViewType == TYPE_FILE_AND_MEDIA_ROW
                ) {
                    fileDescription1.isDownloadError = true
                    notifyItemChanged(i)
                }
            }
        }
    }

    private fun addItems(fileDescriptionList: List<FileDescription?>) {
        list.addAll(getMediaAndFileItems(fileDescriptionList))
    }

    private fun getMediaAndFileItems(fileDescriptionList: List<FileDescription?>): ArrayList<MediaAndFileItem> {
        val result = ArrayList<MediaAndFileItem>()
        if (fileDescriptionList.isEmpty()) return result

        val current = Calendar.getInstance()
        val prev = Calendar.getInstance()
        fileDescriptionList[0]?.let { fd ->
            result.add(DateRow(fd.timeStamp))
            result.add(
                FileAndMediaItem(
                    fd,
                    if (fd.fileUri != null) {
                        getFileName(
                            fd.fileUri!!
                        )
                    } else {
                        ""
                    }
                )
            )
        }
        for (i in 1 until fileDescriptionList.size) {
            fileDescriptionList[i]?.let { fd ->
                current.timeInMillis = fd.timeStamp
                prev.timeInMillis = fileDescriptionList[i - 1]?.timeStamp ?: 0L
                result.add(
                    FileAndMediaItem(
                        fd,
                        if (fd.fileUri != null) {
                            getFileName(
                                fd.fileUri!!
                            )
                        } else {
                            ""
                        }
                    )
                )
                if (current[Calendar.DAY_OF_YEAR] != prev[Calendar.DAY_OF_YEAR]) {
                    result.add(DateRow(fd.timeStamp))
                }
            }
        }

        return result
    }

    private inline fun updateWithDiffUtil(changeItems: () -> Unit) {
        val oldItems = ArrayList(list)
        changeItems()
        val diffCallback = FilesAndMediaDiffUtil(oldItems, list)
        val diff = DiffUtil.calculateDiff(diffCallback)
        diff.dispatchUpdatesTo(this)
    }

    interface OnFileClick {
        /**
         * Описывает реакцию на нажатие
         *
         * @param fileDescription характеристики файла, на котором был
         *     произведен клик
         */
        fun onFileClick(fileDescription: FileDescription?)

        /**
         * Описывает реакцию на начало загрузки файла
         *
         * @param fileDescription характеристики файла, на котором был
         *     произведен клик
         */
        fun onDownloadFileClick(fileDescription: FileDescription?)
    }

    companion object {
        private const val TYPE_DATE_ROW = 1
        private const val TYPE_FILE_AND_MEDIA_ROW = 2
    }
}